tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c

```diff
@@ -0,0 +1,179 @@
+#include <bluetooth/hci.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+#include <misc/byteorder.h>
+
+#include <gatt/hrs.h>
+#include <gatt/bas.h>
+
+static struct bt_conn *default_conn;
+
+/*
+ * Basic connection test:
+ *   We expect a central to connect to us.
+ *
+ *   The thread code is mostly a copy of the peripheral_hr sample device
+ */
+
+#define WAIT_TIME 6 /*seconds*/
+extern enum bst_result_t bst_result;
+
+static void test_con2_init(void)
+{
+	bst_ticker_set_next_tick_absolute(WAIT_TIME*1e6);
+	bst_result = In_progress;
+}
+
+static void test_con2_tick(bs_time_t HW_device_time)
+{
+	/*
+	 * If in WAIT_TIME seconds the testcase did not already pass
+	 * (and finish) we consider it failed
+	 */
+	if (bst_result != Passed) {
+		bst_result = Failed;
+		bs_trace_error_line("test: connect2 failed (not finished "
+				    "after %i seconds)\n", WAIT_TIME);
+	}
+}
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL, 0x0d, 0x18, 0x0f, 0x18, 0x05, 0x18),
+};
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	if (err) {
+		printk("Connection failed (err %u)\n", err);
+	} else {
+		default_conn = bt_conn_ref(conn);
+		printk("Connected\n");
+	}
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	printk("Disconnected (reason %u)\n", reason);
+
+	if (default_conn) {
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void bt_ready(int err)
+{
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	hrs_init(0x01);
+	bas_init();
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		printk("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+}
+
+static void auth_cancel(struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Pairing cancelled: %s\n", addr);
+}
+
+static struct bt_conn_auth_cb auth_cb_display = {
+	.cancel = auth_cancel,
+};
+
+static void test_con2_main(void)
+{
+	int err;
+
+	err = bt_enable(bt_ready);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+	bt_conn_auth_cb_register(&auth_cb_display);
+
+	/* Implement notification. At the moment there is no suitable way
+	 * of starting delayed work so we do it here
+	 */
+	for (int i = 0; i < WAIT_TIME-1; i++) {
+		k_sleep(MSEC_PER_SEC);
+
+		/* Heartrate measurements simulation */
+		hrs_notify();
+
+		/* Battery level simulation */
+		bas_notify();
+	}
+
+	/* We consider it passed */
+	bst_result = Passed;
+	printk("Testcase passed\n");
+
+	/* Wait to be killed */
+	for (;;) {
```

Comment:
Doing it this way (in both sides) the devices will never end the simulation, but you rely on the sim_length passed to the Phy.
(This Phy timeout is 20 simulated seconds now)

If, in this device, all you want is to see it send 1 notification to pass, then you could have it change the test status to passed after the first notification, and just let this device go on sending notifications every second.
You could then let the other side device be the one that ends the simulation (with the same condition it had before).